arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+  auto const quorum_size = _self._config.writeConcern;
+
+  std::vector<std::pair<LogIndex, ParticipantId>> indexes;
+  indexes.reserve(_follower.size());
+  for (auto const& follower : _follower) {
+    // The lastAckedEntry is the last index/term pair that we sent that this
+    // follower acknowledged - means we sent it. And we must not have entries
+    // in our log with a term newer than currentTerm, which could have been
+    // sent to a follower.
+    auto const& lastAckedEntry = follower.lastAckedEntry;
+    TRI_ASSERT(lastAckedEntry.term <= this->_self._currentTerm);
+    // We must never commit log entries for older terms, as these could still be
+    // overwritten later if a leader takes over that holds an entry with the
+    // same index, but with a newer term than that entry has.
+    // For more details and an example see the Raft paper, specifically on
+    // page 9 both subsection "5.4.2 Committing entries from previous terms" and
+    // figure 8.
+    // We may only commit these if we've written an entry in our current term.
+    // This also includes log entries persisted on this server, i.e. our
+    // LocalFollower is no exception.
+    if (lastAckedEntry.term == this->_self._currentTerm) {
+      indexes.emplace_back(lastAckedEntry.index, follower._impl->getParticipantId());
+    } else {
+      LOG_CTX("54869", TRACE, _self._logContext)
+          << "Will ignore follower "
+          << follower._impl->getParticipantId() << " in the following commit index check, as its last log entry (index "
+          << lastAckedEntry.index << ") is of term " << lastAckedEntry.term
+          << ", but we're in term " << _self._currentTerm << ".";
+    }
+  }
+
+  LOG_CTX("a2d04", TRACE, _self._logContext) << "checking commit index on set " << indexes;
+
+  if (quorum_size <= 0 || quorum_size > indexes.size()) {
+    LOG_CTX("24e92", WARN, _self._logContext)
+        << "not enough participants to fulfill quorum size requirement";
+    return {};
+  }
+
+  auto nth = indexes.begin();
+  std::advance(nth, quorum_size - 1);
+
+  std::nth_element(indexes.begin(), nth, indexes.end(),
+                   [](auto& a, auto& b) { return a.first > b.first; });
+
+  auto& [commitIndex, participant] = indexes.at(quorum_size - 1);
+
+  LOG_CTX("6a6c0", TRACE, _self._logContext)
+      << "calculated commit index as " << commitIndex
+      << ", current commit index = " << _commitIndex;
+  TRI_ASSERT(commitIndex >= _commitIndex);
+  if (commitIndex > _commitIndex) {
+    std::vector<ParticipantId> quorum;
+    auto last = indexes.begin();
+    std::advance(last, quorum_size);
+    std::transform(indexes.begin(), last, std::back_inserter(quorum),
+                   [](auto& p) { return p.second; });
+
+    auto const quorum_data =
+        std::make_shared<QuorumData>(commitIndex, _self._currentTerm, std::move(quorum));
+    return updateCommitIndexLeader(commitIndex, quorum_data);
```

Comment:
If `updateCommitIndexLeader`'s signature would be slightly changed, we could move the shared_ptr into it.
```suggestion
    return updateCommitIndexLeader(commitIndex, std::move(quorum_data));
```